PublicMapping/districtbuilder
# Fix excess memory usage on projects query

## Overview

The production database was shutdown by AWS due to excess memory usage, with the potential cause being frequent queries that result in large temporary files being created, as described in [the issue comment](https://github.com/PublicMapping/districtbuilder/issues/1225#issuecomment-1137657992). In order to replicate this, I created ~1500 projects of a little over 50 regions and found similar queries resulting in a large amount of temporary files using `pgbadger` logging. While the temporary files in my local instance were noticeably smaller than those related to the production outage, the query is unique to simplifying the geojson districts on read and of similar frequency. It's assumed that multiple calls from users with a lot of projects of more complex geometries would log temporary file sizes closer to those that were created before the outage. In order to fix this, we decided to denormalize the `project` table to include a `simplified_districts` column, which would be set with the simplified geojson on create rather than read. After implementing, there are no more temporary files created after querying for a user's paginated projects.

### Checklist


### Notes

- The `pgbadger` .html "before" and "after" files are included in this PR for viewing with the intent to take them back out before merge. I created the logs for `pgbadger` by selecting the last page of the user maps screen with 1,467 projects of a little over 50 regions. The query durations are different for the "before" and "after" files, since I tested "before" with a couple other actions to compare (creating a new project, duplicating, navigating site, etc) vs just selecting through to the last page where large temporary files were expected. Below are screenshots of the changes between temporary files creation:
Before implementing changes:
<img width="1131" alt="Screen Shot 2022-06-07 at 8 57 12 AM" src="https://user-images.githubusercontent.com/36374797/172444348-f40dea79-6138-451b-8857-e959aec9296d.png">
<img width="1135" alt="Screen Shot 2022-06-07 at 8 57 02 AM" src="https://user-images.githubusercontent.com/36374797/172444368-9f29af2c-81a5-4665-a6a4-9d3745cfb272.png">

After implementing changes:
<img width="1133" alt="Screen Shot 2022-06-07 at 8 38 24 AM" src="https://user-images.githubusercontent.com/36374797/172444174-edb65f7c-885c-4614-bbb9-64a8436daf3b.png">
<img width="1132" alt="Screen Shot 2022-06-07 at 8 38 36 AM" src="https://user-images.githubusercontent.com/36374797/172444196-03d1b8b4-3d20-4854-a6ef-cab176f5343f.png">

- Since we moved the action of simplifying the geojson from the database back into js, we expect more CPU time dedicated to simplification than before. To evaluate wether it made sense to move this into a worker process, we tested creating various regions, including block level districts, and found simplification didn't take longer than ~26ms for those scenarios. For now it seems worth skipping and reevaluating if it becomes an issue.

## Testing Instructions

- `scripts/update`
- `scripts/server`
- Confirm mini maps display as before
- Create a new project via template and import and confirm each functions as expected and mini map displays
- On updating districts, confirm mini map updates to display new districts
- Create a handful of new projects using `scripts/manage create-random-projects` and confirm new projects have `simplified_districts` field and mini maps display

Closes #1225
